### PR TITLE
[Impeller] fix barriers prior to mip generation.

### PR DESCRIPTION
**Pull request by jonahwilliams**
This fixes the flickering and rendering errors when running on Android seen in https://github.com/flutter/flutter/issues/148412 .

The problem is the barriers proceding mip generation do not account for the input text having been the output of a render pass. Unfortunately no validation errors on this, and no Idea how to test it.

### CODE REVIEW

File: impeller/renderer/backend/vulkan/blit_pass_vk.cc [hunk 1]
```diff
@@ -345,16 +345,18 @@ bool BlitPassVK::OnGenerateMipmapCommand(std::shared_ptr<Texture> texture,
...
+      vk::AccessFlagBits::eTransferWrite |
+          vk::AccessFlagBits::eColorAttachmentWrite,  // src access mask
+      vk::AccessFlagBits::eTransferRead,              // dst access mask
+      src.GetLayout(),                                // old layout
+      vk::ImageLayout::eTransferDstOptimal,           // new layout
+      vk::PipelineStageFlagBits::eTransfer |
+          vk::PipelineStageFlagBits::eColorAttachmentOutput,  // src stage
+      vk::PipelineStageFlagBits::eTransfer,                   // dst stage
+      0u,                                                     // mip level
+      mip_count                                               // mip level count
```
Comment (gaaclarke) [inline impeller/renderer/backend/vulkan/blit_pass_vk.cc:359]:
```suggestion
      /*cmd=*/cmd,
      /*image=*/image,
      /*src_access_mask=*/vk::AccessFlagBits::eTransferWrite |
          vk::AccessFlagBits::eColorAttachmentWrite,
      /*dst_access_mask=*/vk::AccessFlagBits::eTransferRead,
      /*old_layout=*/src.GetLayout(),
      /*new_layout=*/vk::ImageLayout::eTransferDstOptimal,
      /*src_stage=*/vk::PipelineStageFlagBits::eTransfer |
          vk::PipelineStageFlagBits::eColorAttachmentOutput,
      /*dst_stage=*/vk::PipelineStageFlagBits::eTransfer,
      /*base_mip_level=*/0u,
      /*mip_level_count=*/mip_count
```

**Review by gaaclarke — COMMENTED** [review]
We do have tests that get close to this in blit_pass_unittests.cc.  None of them hit this code at all.  Can you add a test that executes `BlitPassVK::OnGenerateMipmapCommand` to make sure it doesn't crash or create validation errors?  That's better than nothing.

**Comment by jonahwilliams** [comment]
I can definitely add a test case that hits this path. That said, I did not observe any validation errors locally unfortunately

**Comment by gaaclarke** [comment]
> I can definitely add a test case that hits this path. That said, I did not observe any validation errors locally unfortunately

Yea, the test won't help us know that we've fixed the issue, but it will help us know that we didn't introduce a new one.  I was hoping maybe the process of writing a non-perfect test might make the path to a better test more clear.

**Review by chinmaygarde — APPROVED** [review]
The sync validation only catches errors due such as write-after-write which can be trivially checked to be incorrect. This one is an issue with insufficient synchronization and harder to detect. For instance, we could have had a compute stage writing to this texture before before blitting from it. Or perhaps we had another form of synchronization like a semaphore.

Like you idea on tracking layouts at the queue level, perhaps we can track write and reads at the per-resource per-queue level too and automatically insert barriers? 

LGTM modulo Aarons comments.

**Comment by jonahwilliams** [comment]
Longer term, yeah - automatic barrier determination is a must - especially for something like flutter GPU. That said, its a big lift to do so efficiently

**Comment by jonahwilliams** [comment]
I checked and all of the backdrop blurs in aiks_blur_unttests will hit this case. The bug happens when a render pass texture has mips generated

**Comment by gaaclarke** [comment]
> The barriers _are_ currently in a bad state

Could we get that bug to show up in the golden output then by drawing at a particular mip level?  Or is it non deterministic so we couldn't know for sure we grabbed bad output?

**Comment by jonahwilliams** [comment]
Unfortunately its a synchronization issue/race, so I don't think we can guarantee it shows up

**Comment by gaaclarke** [comment]
> Unfortunately its a synchronization issue/race, so I don't think we can guarantee it shows up

Ahh yea, we'd probably have to make a test that just pounds on drawing and check that the output is the same each frame, so not really a golden test.  Otherwise, a test that just tests the flags on a mocked up vulkan isn't so helpful.  We don't really have anything that does that yet that I know of.  Do you think we could reproduce the problem like that and would it be worth investing in?  This bug is so subtle I could easily see this getting broken again at some point and not noticing it until much later, but investing in a potentially flaky test isn't an obvious slam dunk either.

**Comment by jonahwilliams** [comment]
if all of our blurs start flickering, I think we just have to hope that we notice. I do not even know if its possible to reproduce in swiftshader. Its possible that the implicit synchronization from running everything on the CPU causes it to "just work".

**Review by gaaclarke — APPROVED** [review]
test exception: The new logic is invariant so a test that mocks out vulkan isn't that informative.  The code is executed as part of the blur tests, so it has some coverage to make sure it isn't completely wrong.  The error manifests because of a synchronization problem that isn't caught by the validation layers.  We are uncertain if pounding on drawing could even reproduce the problem with our swiftshader test harness.

**Pull request merged**